### PR TITLE
OEUI-285: Validate if drug order is active

### PR DESCRIPTION
**Pull request by dessHub**
### **JIRA TICKET NAME**
[OEUI-285: Validate if drug order is active](https://issues.openmrs.org/browse/OEUI-285) 

### **Summary**
- I wrote a function to check if there is an active drug order before adding the drug order in the draft list. 

## I have checked that on this Pull request

- [X] I can sucessfully create Drug orders
- [X] I can sucessfully create Lab orders
- [X] I can sucessfully search for drug orders

### CODE REVIEW

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 1]
```diff
@@ -95,48 +104,52 @@ export class AddForm extends React.Component {
...
-    });
-    this.props.selectDrugSuccess('');
-    this.clearDrugForms();
-    this.props.clearSearchField();
-    if (this.props.activity === 'DRAFT_ORDER_EDIT') {
-      this.props.setSelectedOrder({
-        currentOrderType: this.props.currentOrderType,
-        selectedOrder: null,
-        activity: null,
+    if (this.checkIfDrugExist(this.props.drugName) < 1) {
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Will work on it right away

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 2]
```diff
@@ -95,48 +107,52 @@ export class AddForm extends React.Component {
...
+          dose,
+          dosingUnit,
+          frequency,
+          route,
+          duration,
+          durationUnit,
+          dispensingUnit,
+          dispensingQuantity,
+          reason,
+          previousOrder: this.state.previousOrder,
```
Comment (Rhotimee) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Hey @dessHub, It'll be nice if you could destructure this.state and this.props.

File: webpack.config.js [hunk 3]
```diff
@@ -25,7 +25,7 @@ const autoprefixer = require('autoprefixer');
 
 const env = process.env.NODE_ENV;
 
-const THIS_APP_ID = 'orderentry';
+const THIS_APP_ID = 'openmrs-owa-orderentry';
```
Comment (Rhotimee) [inline webpack.config.js]: @dessHub, why did you change the name to `openmrs-owa-orderentry` ?

Comment (dessHub) [reply webpack.config.js]: I can't remember changing it, but I guess it was a mistake. Will take care of it.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 4]
```diff
@@ -81,6 +83,13 @@ export class AddForm extends React.Component {
     this.setState({ formType: formType.trim() });
   }
 
+  checkIfDrugExist = (drug) => {
+    const { filteredOrders } = this.props.fetchOrdersReducer;
+    const drugOrders = filteredOrders.filter(order => order.type === 'drugorder');
+    const findDrug = drugOrders.filter(order => order.drug.display === drug);
+    return findDrug;
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I have taken of it. Thank you for the reviews.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 2 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 5]
```diff
@@ -81,6 +83,16 @@ export class AddForm extends React.Component {
     this.setState({ formType: formType.trim() });
   }
 
+  checkIfDrugExist = (drug) => {
```
Comment (dkayiwa) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Check if drug exists or if drug has an active order?

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 6]
```diff
@@ -81,6 +83,16 @@ export class AddForm extends React.Component {
     this.setState({ formType: formType.trim() });
   }
 
+  checkIfDrugHasActiveOrder = (drug) => {
+    const { filteredOrders } = this.props.fetchOrdersReducer;
+    const drugOrders = filteredOrders.filter(order => order.type === 'drugorder');
+    const findDrug = drugOrders.filter(order => order.drug.display === drug);
```
Comment (dondrzzy) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Please use clearer variable names.
1: Equating a `display` name to a drug looks like a `contradiction`. Use displayName or something similar instead of `drug` as a parameter.
2: Variable name `findDrug` sounds like a function and not a drug or object. Find a name that better describes a matching orders that were found. My guess it has to suit an array since you used a filter.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 7]
```diff
@@ -81,6 +83,16 @@ export class AddForm extends React.Component {
     this.setState({ formType: formType.trim() });
   }
 
+  checkIfDrugHasActiveOrder = (drug) => {
+    const { filteredOrders } = this.props.fetchOrdersReducer;
```
Comment (dondrzzy) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: In `componentDidMount`, `fetchOrders` action returns only 10 orders, which is the default limit meaning your filtered orders are a proportion of all relevant orders in the database. If you are going to look through these 10 orders to determine if a drug name exists then I think this fails in case the drug was not returned among the 10 orders. Does this make sense to you? I would rather you find a way to check all orders rather than those 10 returned when you mount the component.

Comment (dondrzzy) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Please use destructuring props assignment.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 6 above)
Comment (dondrzzy) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Why are you using the`drug display` name to determine whether the drug exists? I think there are unique Identifiers that serve this purpose better.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 8]
```diff
@@ -81,6 +83,16 @@ export class AddForm extends React.Component {
...
+    const drugOrders = filteredOrders.filter(order => order.type === 'drugorder');
+    const findDrug = drugOrders.filter(order => order.drug.display === drug);
+    if (findDrug.length > 0) {
```
Comment (dondrzzy) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I think since a user cant order for the same drug more than once,  exactly equal to 1 would be more discrete. `>0` takes on a situation where the same order is active 100 times.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 9]
```diff
@@ -95,48 +107,58 @@ export class AddForm extends React.Component {
...
+      this.props.clearSearchField();
+      if (this.props.activity === 'DRAFT_ORDER_EDIT') {
+        this.props.setSelectedOrder({
+          currentOrderType,
+          selectedOrder: null,
+          activity: null,
+        });
+      }
+    } else {
+      errorToast(`${this.props.drugName} Order is active.`);
```
Comment (dondrzzy) [inline app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I think telling a user an order is active as a message might not be so direct. Maybe telling them that order already exists might be more descriptive. What do you think? Also, a small`o` would yield a better sentence. For example:`Paracetamol order already exists`.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: This makes sense. Will make the changes.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 10]
```diff
@@ -95,48 +107,58 @@ export class AddForm extends React.Component {
...
-    });
-    this.props.selectDrugSuccess('');
-    this.clearDrugForms();
-    this.props.clearSearchField();
-    if (this.props.activity === 'DRAFT_ORDER_EDIT') {
-      this.props.setSelectedOrder({
-        currentOrderType: this.props.currentOrderType,
-        selectedOrder: null,
-        activity: null,
+    if (!this.checkIfDrugExist(this.props.drugName)) {
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I agree with you. Since drug name is unique, I thought its fine to use it as an identifier. Will update it to use Uuid.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 11]
```diff
@@ -424,6 +450,13 @@ AddForm.propTypes = {
     errorMessage: PropTypes.string,
     addedOrder: PropTypes.object,
   }).isRequired,
+  fetchOrdersReducer: PropTypes.shape({
+    filteredOrders: PropTypes.array,
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I thought it is the objects which take on a shape only. Will read more on it and update accordingly.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 6 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Thanks for the feedback, will work on your suggestions.

Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Will refactor to use uuid

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 8 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: This makes sense. Will update it

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 9 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Will make the changes.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 7 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Will check on this. I had missed the limit parameter. I can add a function to ferchOrders.js action file to fetch all orders from the API.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 12]
```diff
@@ -95,48 +107,58 @@ export class AddForm extends React.Component {
...
+    if (!this.checkIfDrugExist(this.props.drugName)) {
+      const {
+        previousOrder, formType, action, draftOrder,
+      } = this.state;
+      const {
+        careSetting, drugUuid, session, draftOrders, orderNumber, currentOrderType,
+      } = this.props;
+      this.setState({
+        draftOrder: {
+          drugName: this.props.drugName,
```
[the remark this answers is not in the document]
Comment (dondrzzy) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Oh! @dessHub, that makes sense. Only refactor what you added but I think you can go ahead and create a different PR to properly format the file.

[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Awesome. I will do that.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 7 above)
Comment (dondrzzy) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Instead of getting all orders everytime a form is mounted, I think there is already a backend route to check for the existence of an order.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: This ticket will take care of this.

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx [hunk 13]
```diff
@@ -95,48 +107,58 @@ export class AddForm extends React.Component {
...
+          action: this.state.action,
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/index.jsx [hunk 14]
```diff
@@ -231,4 +231,4 @@ export default connect(
     deleteDraftOrder,
     selectDrugSuccess,
   },
-)(SearchAndAddOrder);
+)(SearchAndAddOrder);
```
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/index.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: I have taken care of this. I dont know why it was failing.

Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 7 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 6 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 8 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done

File: app/js/components/drugOrderEntry/addForm/AddForm.jsx (hunk 9 above)
Comment (dessHub) [reply app/js/components/drugOrderEntry/addForm/AddForm.jsx]: Done
